pkg/kn/commands/plugin/plugin_handler.go

```diff
@@ -0,0 +1,109 @@
+// Copyright © 2019 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
+)
+
+// PluginHandler is capable of parsing command line arguments
+// and performing executable filename lookups to search
+// for valid plugin files, and execute found plugins.
+type PluginHandler interface {
+	// exists at the given filename, or a boolean false.
+	// Lookup will iterate over a list of given prefixes
+	// in order to recognize valid plugin filenames.
+	// The first filepath to match a prefix is returned.
+	Lookup(filename string) (string, bool)
```

Comment:
This doesn't seem like it takes a filename, it takes a concatenated-with-dash-subcommand-aggregation. That's a messy interface, I think we can do better.

Reply:
That's what `kubectl` plugins do. I think we should be careful in not breaking existing agreed upon interfaces. We'd just be re-inventing something slightly different IMO.

Reply:
The arg name at least shouldn't be misleading.

Reply:
@rhuss and @sixolet here's how it's working:

```bash
➜  client git:(kn-plugins4) cat /usr/local/bin/kn-foo_bar
#!/bin/bash

echo "Hello Knative, I'm a Kn plugin"
➜  client git:(kn-plugins4) cat /usr/local/bin/kn-foo-bar
#!/bin/bash

echo "Hello Knative, I'm a Kn plugin"
➜  client git:(kn-plugins4) ./kn --lookup-plugins-in-path=true plugin list
Using kn config file: /Users/maximilien/.kn/config.yaml
The following compatible plugins are available, using options:
  - plugins dir: '~/.kn/plugins'
  - lookup plugins in path: 'true'

/Users/maximilien/.kn/plugins/kn-hello
/usr/local/bin/kn-foo-bar
/usr/local/bin/kn-foo_bar
/usr/local/bin/kn-hello
  - warning: /usr/local/bin/kn-hello is overshadowed by a similarly named plugin: /Users/maximilien/.kn/plugins/kn-hello
/usr/local/bin/kn-kn
/usr/local/bin/kn-knctl

error: one plugin warning was found

➜  client git:(kn-plugins4) ./kn --lookup-plugins-in-path=true foo-bar
Hello Knative, I'm a Kn plugin
➜  client git:(kn-plugins4) ./kn --lookup-plugins-in-path=true foo bar
Hello Knative, I'm a Kn plugin
```